File: Tree/BinarySearchTree/BinarySearchTree.cpp
#include <iostream>

class TreeNode {
public:
    int key;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int k) : key(k), left(nullptr), right(nullptr) {}
};

class BinarySearchTree {
private:
    TreeNode* root;

    TreeNode* insert(TreeNode* node, int key) {
        if (node == nullptr)
            return new TreeNode(key);

        if (key < node->key)
            node->left = insert(node->left, key);
        else if (key > node->key)
            node->right = insert(node->right, key);

        return node;
    }

    TreeNode* search(TreeNode* node, int key) {
        if (node == nullptr || node->key == key)
            return node;
        if (key < node->key)
            return search(node->left, key);
        return search(node->right, key);
    }

    void inorderTraversal(TreeNode* node) {
        if (node != nullptr) {
            inorderTraversal(node->left);
            std::cout << node->key << " ";
            inorderTraversal(node->right);
        }
    }

public:
    BinarySearchTree() : root(nullptr) {}

    void insert(int key) {
        root = insert(root, key);
    }

    bool search(int key) {
        return search(root, key) != nullptr;
    }

    void printInOrderTraversal() {
        inorderTraversal(root);
    }
};

int main() {
    BinarySearchTree bst;
    bst.insert(10);
    bst.insert(5);
    bst.insert(15);
    bst.insert(3);
    bst.insert(7);

    std::cout << "Inorder traversal: ";
    bst.printInOrderTraversal();
    std::cout << std::endl;

    bool searchResult = bst.search(10);
    if (searchResult)
        std::cout << "Node with key 10 found." << std::endl;
    else
        std::cout << "Node with key 10 not found." << std::endl;

    return 0;
}

File: Tree/BinarySearchTree/BinarySearchTree.cs
using System;

public class TreeNode
{
    public int Key { get; set; }
    public TreeNode Left { get; set; }
    public TreeNode Right { get; set; }

    public TreeNode(int key)
    {
        Key = key;
        Left = null;
        Right = null;
    }
}

public class BinarySearchTree
{
    private TreeNode root;

    private TreeNode Insert(TreeNode node, int key)
    {
        if (node == null)
            return new TreeNode(key);

        if (key < node.Key)
            node.Left = Insert(node.Left, key);
        else if (key > node.Key)
            node.Right = Insert(node.Right, key);

        return node;
    }

    private TreeNode Search(TreeNode node, int key)
    {
        if (node == null || node.Key == key)
            return node;
        if (key < node.Key)
            return Search(node.Left, key);
        return Search(node.Right, key);
    }

    private void InOrderTraversal(TreeNode node)
    {
        if (node != null)
        {
            InOrderTraversal(node.Left);
            Console.Write(node.Key + " ");
            InOrderTraversal(node.Right);
        }
    }

    public void Insert(int key)
    {
        root = Insert(root, key);
    }

    public bool Search(int key)
    {
        return Search(root, key) != null;
    }

    public void PrintInOrderTraversal()
    {
        InOrderTraversal(root);
    }
}

class Program
{
    static void Main()
    {
        BinarySearchTree bst = new BinarySearchTree();
        bst.Insert(10);
        bst.Insert(5);
        bst.Insert(15);
        bst.Insert(3);
        bst.Insert(7);

        Console.Write("Inorder traversal: ");
        bst.PrintInOrderTraversal();
        Console.WriteLine();

        bool searchResult = bst.Search(10);
        if (searchResult)
            Console.WriteLine("Node with key 10 found.");
        else
            Console.WriteLine("Node with key 10 not found.");
    }
}